application-urlshortener-default/src/main/java/com/xwiki/urlshortener/internal/URLShortenerResourceReferenceHandler.java

```diff
@@ -124,6 +127,20 @@ private List<?> getURLShortenerObjectWithID(URLShortenerResourceReference resour
         if (!resourceReference.getWikiId().isEmpty()) {
             query = query.setWiki(resourceReference.getWikiId());
         }
-        return query.execute();
+
+        List<?> queryResults = query.execute();
+
+        if (queryResults.isEmpty()) {
+            // If no short url is found on the given subwiki, try to find the pageID in all subwikis.
+            // Note that the query is very slow when solr is reindexing.
+            statement = "property.URLShortener.Code.URLShortenerClass.pageID:" + resourceReference.getPageId();
+            query = this.queryManager.createQuery(statement, "solr");
+            QueryResponse response = (QueryResponse) query.execute().get(0);
+            queryResults = response.getResults().stream().map(
+                    // Get document reference without the locale part (which would lead to an inexistent page).
+                    (SolrDocument doc) -> doc.getFieldValue("wiki").toString() + ":" + doc.getFieldValue("fullname"))
```

Comment:
Couldn't you use a `EntityReferenceResolver<SolrDocument>` https://github.com/xwikisas/application-urlshortener/blob/main/application-urlshortener-default/src/main/java/com/xwiki/urlshortener/internal/rest/DefaultURLShortenerResource.java#L100 ?